ServerCore/Pages/Teams/MergeInto.cshtml.cs

```diff
@@ -45,40 +47,42 @@ public async Task<IActionResult> OnPostAsync(int teamId)
             Team = await _context.Teams.FindAsync(teamId);
             var mergeIntoTeam = await _context.Teams.FindAsync(MergeIntoID);
 
-            if (Team == null || mergeIntoTeam == null)
+            if (Team == null || mergeIntoTeam == null || Team.Event != Event || mergeIntoTeam.Event != Event)
             {
                 return NotFound();
             }
 
             var members = await _context.TeamMembers.Where(tm => tm.Team.ID == teamId).ToListAsync();
-            var memberEmails = members.Select(m => m.Member.Email).ToList();
+            var memberEmails = await _context.TeamMembers.Where(tm => tm.Team.ID == teamId).Select(tm => tm.Member.Email).ToListAsync();
             var mergeIntoMemberEmails = await _context.TeamMembers.Where(tm => tm.Team.ID == MergeIntoID).Select(m => m.Member.Email).ToListAsync();
 
             var states = await PuzzleStateHelper.GetSparseQuery(_context, Team.Event, null, Team).ToListAsync();
             var mergeIntoStates = await PuzzleStateHelper.GetSparseQuery(_context, Team.Event, null, mergeIntoTeam).ToDictionaryAsync(s => s.PuzzleID);
 
-            // copy all the team members over
-            foreach (var member in members)
+            using (var transaction = await _context.Database.BeginTransactionAsync(IsolationLevel.Serializable))
```

Comment:
This should happen before the queries above so that their results can't change during the merge.